### PR TITLE
feat(client): shortcuts legend modal available on pressing the `?` key

### LINKED ISSUE #36841: Shortcuts legend modal available on pressing the `?` key.
Shortcuts legend modal available on pressing the `?` key.

Youtube's modal:
https://user-images.githubusercontent.com/985197/64934197-bd627980-d80e-11e9-86e3-961745d0c665.png

### PR DESCRIPTION
**Pull request by raditotev**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #36841

This is not the final version, therefore I open this as a draft PR.
To complete this task I would need more information, namely:
* ~For some reason the texts I've added in the `translations.json` files are not available inside  the component. This is the reason I've left the hard coded text and the attempt to use the text from the translations.~ [Comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/45530#issuecomment-1101224993)
    ```jsx
    // client/src/templates/Challenges/components/shortcuts-modal.tsx:49
  
    <Modal.Title className='text-center'>
      Keyboard shortcuts <br />
      {t('shortcuts.title')}
    </Modal.Title>
    ```
* ~I have added 3 shortcuts (previous exercise, next exercise and close modal). Are there any more?~ (https://github.com/freeCodeCamp/freeCodeCamp/issues/36841#issuecomment-933310078)
*  ~What should the data to google analytics look like? This is what I have now:~ [Comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/45530#issuecomment-1103754231)
   `executeGA({ type: 'modal', data: '/shortcuts-modal' });`
* ~I had to disable one eslint rule in order to capture user key strokes:~ [Comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/45530#issuecomment-1101224993)
  ```tsx
    // client/src/components/layouts/learn.tsx:99
    {/* eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions */}
    <main id='learn-app-wrapper' onKeyPress={keyPressHandler}>
      {children}
    </main>
  ```
  ~I am not absolutely sure how this will affect the accessibility. I would like to hear your thoughts on this. Or there might be another way to capture key strokes I have't thought of.~
* ~styling is another thing where some feedback will be appreciated~

### CODE REVIEW

**Comment by raisedadead** [comment]
Hi @raditotev 

Thanks for working on this feature. 

> This is not the final version, therefore I open this as a draft PR.

Contrary to the belief, draft PRs as typically work in progress by the OP and are not ready for any review. I believe we missed getting back to you because of this. Please accept my apologies.

@ojeytonwilliams @bbsmooth and @ahmadabdolsaheb, I am tagging you all for helping answer the queries relevant to your areas of expertise. Can you please guide us here?

Thanks, everyone!

**Comment by ojeytonwilliams** [comment]
Hi there @raditotev.  For the translations, everything's fine.  The reason you saw the key rather than the translation is that we haven't made Gatsby watch the i18n files and so it doesn't update until you clear the cache and rebuild. `npm run clean:client` should do the trick.

As for the global keypress, it should be fine to use the [GlobalHotKeys](https://github.com/greena13/react-hotkeys#globalhotkeys-component) component in the learn layout, rather than trying to attach the listener to `main`.

**Comment by bbsmooth** [comment]
@raisedadead Thanks for tagging me in.

There are several accessibility issues here. The first one we are going to run into is [WCAG SC 2.1.4.](https://www.w3.org/WAI/WCAG21/Understanding/character-key-shortcuts.html). Using single key shortcuts cause issues for screen reader and voice control users. If we wanted to continue to use single key shortcuts then the two options we have are to either provide a mechanism to turn them off or provide a mechanism to allow the user to remap them. Personally, I think it might be best if we add a modifier key (or two) to them, just like we do with `Ctrl+Enter`.

Another issue is that the "?" key works while focus is in the editor and will open the modal, which is probably not a good idea since the user probably does not expect nor want the modal to open up if the "?" key is used while typing. Also, FF will type the "?" in the editor but Chrome won't type the "?" in the editor. So we have two different behaviors for the same keyboard shortcut. But really, if we are going to keep the single key "?" shortcut then we should probably not allow it to function in the editor.

The content in the modal looks like a data table to me and should probably be marked up as such. I think doing this would make it easier for screen reader users to access the info since they can use the built in table keyboard shortcuts to navigate the table and each value in the table will be associated with a proper heading. Right now there really is no programmatic relationship defined between the function and the key.

Also, because there is so much content in the modal, I would not have the initial focus set on the dialog itself, which causes the screen reader to automatically start reading everything from the top down. Rather, I would have the focus set on the close button in the upper right and then screen reader users can manually navigate the table. 

We also need to give the dialog an accessible name by adding the `aria-labelledby` attribute to the `<div>` with `role="dialog"` and have it point to the main heading ("Keybord shortcuts"). That heading will need to have an `id` so we can point to it.

I'll stop here. There may be a few other minor issues but I think I gave us enough to talk about for now.

**Comment by raisedadead** [comment]
Thanks for your feedback @bbsmooth 

> ...I think it might be best if we add a modifier key...

I am in favor of doing this and I think others would be as well.

> ...the "?" key works while focus is in the editor and will open the modal...

You are correct, @raditotev we should remove/disable the shortcut while the focus is in the editor. We do not need any triggers while coding.

> The content in the modal looks like a data table to me and should probably be marked up as such...
and
> I would have the focus set on the close button in the upper right and then screen reader users can manually navigate the table.

These are really nice catches, we would not have thought about it. That said, I think to limit the scope can we do a follow-up PR to put the content in a table as per your recommendation right after we get this in? Anyways I will defer to @raditotev on how we want to proceed.

In any case, thanks a lot for your thoughts on this. 

...

I think we have much of the initial questions from OP sorted. @ahmadabdolsaheb please advise on the query about GA and I believe we can have more progress on the changes.

**Comment by raisedadead** [comment]
I did not realize this prior to your comment, but GitHub, and Gmail, all use `Shift + ?` to show the shortcuts in a modal-like interface. We should definitely use this combination.

**Comment by bbsmooth** [comment]
Definitely, "?" can stay for opening up the modal since it is `Shift +  /` which uses a modifier key and thus is not a single character key shortcut. But that won't work with the letter keys since SC 2.1.4 considers both lowercase and uppercase letters to be a single character.

**Comment by bbsmooth** [comment]
After looking into this a bit more I need to retract what I just said about the "?" key. SC 2.1.4 applies to the character being produced by the keyboard, not the actual keys used to produce the character. We can continue to use "?" to open this modal but since it is just a one character shortcut then we need to allow the user to ability to either turn it off or remap it.

**Comment by bbsmooth** [comment]
I understand your argument for voice control readers, but this also applies to keyboard users as well and I think we run into a problem there.

Is it possible to detect `Shift + ?` from a keyboard event? The only way I can produce `?` on my keyboard is to use `Shift + /`, so `shiftKey` is always going to be `true` in the event when typing "?". In other words, listening for just `?` or `Shift + ?` is the same thing, you can't have one without the other (at least with a typical US keyboard layout). I think `Shift` can only be considered a modifier for keys that don't produce a printable character, such as the arrow keys or F keys.

**Comment by ojeytonwilliams** [comment]
Nope, it's not possible (at least on my keyboard) to get a question mark without pressing shift.  However, from a keyboard user's perspective that's fine: they can't accidentally trigger the modal without having a second key pressed. 

Speech input could be an issue, since it's conceivable that they would generate a `?` by simulating a keyboard event with `shiftKey: true` and `code: 'Slash'`.  I think that's unlikely, but the speech input software I've found doesn't generate keyboard events at all, so I've not been able to test it.

**Comment by ahmadabdolsaheb** [comment]
The whole `executeGA` could be removed as we don't need lot have analytics for this modal.

**Comment by raditotev** [comment]
Thank you all for the input. @raisedadead I have tried to add focus to the modal as suggested by @bbsmooth but I get an error:
```
Type '{ children: Element[]; autoFocus: true; }' is not assignable to type 'DetailedHTMLProps<TableHTMLAttributes<HTMLTableElement>, HTMLTableElement>'.
  Property 'autoFocus' does not exist on type 'DetailedHTMLProps<TableHTMLAttributes<HTMLTableElement>, HTMLTableElement>'.
```
Also, I saw later you have suggested to change the list items in the body with table in a different task but I have already done it by the time I've read your comment.

**Comment by raditotev** [comment]
I saw I have left a couple unused styles. As I suspect there will be a few more changes I won't push the changes just yet. I would like to hear if there are any suggestions regarding the overall style of the modal:

Shortcuts modal

**Comment by raisedadead** [comment]
@raditotev 

This looks great! Let's wait for feedback from others.

--

> ...I have tried to add focus to the modal as suggested...

OK – do not worry about it, for now, we can certainly look at it later.

> Also, I saw later you have suggested to change the list items in the body with table in a different task but I have already done it by the time I've read your comment.

Of course, no worries. I was trying to reduce the work :)

**Comment by bbsmooth** [comment]
> However, from a keyboard user's perspective that's fine: they can't accidentally trigger the modal without having a second key pressed.

I'm not sure we can make the assumption that someone will always be pressing two keys to produce a "?". There may be keyboard layouts where the `Shift` key isn't required. Also, there may be non-keyboard interfaces that simulate keyboard events and the user can generate the "?" directly with a single "keypress" or action (e.g. virtual keyboards, touch screens).

SC 2.1.4 is ultimately about the character produced by the keyboard, not the number of keys it takes to produce a character. If we want to offer the "?" shortcut for the modal and be in compliance (which I think we do since this is a level A criterion) then we will need to either provide a way to turn of this shortcut or a method for remapping it.

**Comment by raisedadead** [comment]
@bbsmooth 

I understand where you are getting at. I am just not sure why turning off the feature altogether when we are focused within the editor not be sufficient? I do not see why would the use need the `?` outside the editor at all. There is no other text-input interface that would be needing the key.

**Comment by bbsmooth** [comment]
> I do not see why would the user would need the `?` outside the editor at all.

Just to clarify, this isn't just about the `?` key. We would also need to be able to turn off or remap all of the single character shortcut keys (`n`, `p`, `r`, `e`).  The user doesn't need any of these keys outside of the editor, but we are listening for them when the focus is not in the editor in order to implement the shortcuts, and someone using a keyboard (or simulated keyboard interface) can accidentally press one of those keys while their focus is outside of the editor and trigger the shortcut. The intent of SC 2.1.4 is to keep accidents like this from happening.

**Comment by ojeytonwilliams** [comment]
> I'm not sure we can make the assumption that someone will always be pressing two keys to produce a "?". 

Ah, I think I see what we're disagreeing about!  So, to reiterate: we can make shift being pressed a requirement for the hotkey to trigger.  It will not trigger unless the event comes with the `key: '?'` and `shiftKey: true`.

Re: virtual keyboards and touch screens, I couldn't get those to appear when not in the editor, so I think we're good on that front.

**Comment by ojeytonwilliams** [comment]
> We would also need to be able to turn off or remap all of the single character shortcut keys (n, p, r, e). 

Yep, those were definitely a mistake.  I'm not sure *what* we want for those, I agree entirely that those need changing.

**Comment by raisedadead** [comment]
I am onboard about adding a setting that will turn off *all* the shortcuts. The idea is these will be used by power users. 

Something along the lines of (pseudo-code):

```js

if(isShortcutKeysDisabled === false && keypress === 'matchedcombination') {
	console.log('do it');
}

```

The negative flag `isShortcutKeysDisabled` is intentional to enable shortcuts by default for all users. The modal can then have a link to settings or a button to disable all shortcuts.

Is this something agreeable?

**Comment by bbsmooth** [comment]
> It will not trigger unless the event comes with the `key: '?'` and `shiftKey: true`.

But what if someone has a keyboard and they don't need `Shift` to type a "?", and if they do press `Shift` the "?" key produces a different character? The `Shift` key is not a regular modifier key like `Ctrl` and `Alt`. For keys that produce printable characters it acts as a toggle between two characters. Thus, you can't require `Shift` for printable character keys.

> Re: virtual keyboards and touch screens, I couldn't get those to appear when not in the editor, so I think we're good on that front.

Someone who can't use a physical keyboard could be using a variety of touch screen devices for everything they do to interface with a computer, so for them it is their only way of interfacing with a computer and thus the virtual keyboard is always there regardless of whether focus is in the editor.

You all are asking very good questions. I asked these same questions when I was researching this SC as well. And I completely understand that it seems intuitive that `Shift + ?` would pass this requirement (I originally thought it did as well). But for all the reasons I've given above (and probably more) it doesn't fulfill SC 2.1.4.

@raisedadead This is exactly the sort of solution I was thinking of as well and I'm 100% on board with you.

**Comment by raisedadead** [comment]
OK - thanks for confirming all this everyone. Now to keep the work moving I think @raditotev you should clean this up for a final review (that is us pulling the code, playing with it, etc.)

The modal styles look decent to me. I would drop the subtitle saying "List of Shortcuts" and that should be it.

As for the actionable items:

1. Add a setting to disable the shortcuts - This is fairly straight but should be complex enough to require a follow-up PR. However, @raditotev if you are interested we could get it in here. We can help guide you with the changes.

2. Focus on the close button within the modal - This one needs closer inspection. I believe we will have an idea when we review it once the code is cleaned up and ready.

**Comment by raditotev** [comment]
I definitely learned some new things in this discussion 😄 
@raisedadead I've pushed the latest changes including the focus on the close button. I was using cameCase the first time I've tried and that is the reason it wasn't working. Correct form is `autofocus`.

> Add a setting to disable the shortcuts - This is fairly straight but should be complex enough to require a follow-up PR. ? However, @raditotev if you are interested we could get it in here. We can help guide you with the changes.
I would like to make those changes and I think they belong in this PR.

There is one more thing I believe is missing at the moment and is logical to be in the scope of this work. We don't really communicate with the user the existence of this modal. What in your opinion is the best place to give a hint?

**Comment by raisedadead** [comment]
Yes, we all did – Bruce has been doing phenomenal work in helping improve our a11y best practices.

Awesome – Thanks for getting the focus working.

For the settings changes here is what I think we would want (off the top of my head):

1. Add a checkbox in the modal footer area, that says something along the lines of "Disable all keyboard shortcuts, you can enable them again on your settings page".

3. On the API side, this flag should correspond to a new boolean property in the user settings called `keyboardshorcuts`.

   Here is an example of how we use `theme` and `sound` for example.

   https://github.com/freeCodeCamp/freeCodeCamp/blob/766ae44972735849e08cfb20b66f5e583ca9369f/api-server/src/common/models/user.json#L301-L308 

   You would ideally want to add this new property alongside these existing ones and follow their usage to create similar logic in code for the shortcuts flag.

4. Next on the client-side, you will be needing to add a toggle similar to `theme`, etc. on the `/settings` page, again you can follow the usage for `sound` for inspiration.

I may have missed things, but please reach out if you find yourself stuck.

**Comment by bbsmooth** [comment]
Hey @raditotev this is looking real good. I have just a few requests for changes.

The `aria-labelledby` attribute takes an `id` not a string, so we'll need to add an `id` to the `<h4>` heading and then set `aria-labelledby` to that `id`. This will work nicely since the heading will reflect the current language selected and then that will be the accessible name for the dialog.

Speaking of current languages, we should use the `closeLabel` property on `<Modal.Header>` to set the accessible name on the close button to the translated 'buttons.close'.

Concerning the heading, since it creates a new context (in other words, it is like a separate page) I think it might be best if it was an `<h1>` instead of an `<h4>`. It's not the end of the world if this can't be done (it's not technically a WCAG failure) but screen reader users might think they've missed something if we jump right to `<h4>`.

The `autofocus` attribute` doesn't seem to be working for me (I think it needs to be `autoFocus` (capital "F"). Regardless, I think we want the focus to be on the first close button at the top right instead of the close button at the bottom. That way the focus is set right above the table and then a screen reader user can start easily navigating through the table from the top down. I'm not seeing anything in the react-bootstrap docs to do this easily. I was able to make it work by using the `onEntered` property on the `<Dialog>` to call a function that sets the focus on the button.

I'm also noticing that the parent `<div>` of the modal also has `role=dialog`. We should definitely get rid of that if possible.

I still need to do a little more research into screen readers and single character shortcut keys. By default screen readers are in browse (or document) mode when the user isn't interacting with a form input. Browse mode intercepts keyboard presses before it passes them onto the browser, so a screen reader user can't use any of these shortcuts in that mode because "n" and "p" and "?" are already shortcut commands for the screen reader and thus those key presses will never get to the browser. You can switch to forms mode at anytime though, and in forms mode the screen reader won't intercept the keyboard presses and will instead pass them on directly to the browser. So a screen reader user can switch to forms mode real quick in order to use our shortcuts. I just want to make sure this is something screen reader users expect to do or if there is some other accepted best practice they expect to use.

**Comment by raditotev** [comment]
@bbsmooth I am working on making the requested changes but I have couple questions:
* I changes the `autofocus` input with `autoFocus` and tried to use it I get the following warning:
  > The autoFocus prop should not be used, as it can reduce usability and accessibility for users.eslint[jsx-a11y/no-autofocus](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/tree/HEAD/docs/rules/no-autofocus.md)
* Would it be a problem if I use `autoFocus` in the title element `Modal.Title` or if that doesn't work in the `table`

@raisedadead I don't seem to be able to track from where the `role=dialog` in the parent div comes from. Do you have any tips in that regard?
image

**Comment by bbsmooth** [comment]
@raisedadead @ojeytonwilliams Question for everyone. Do we have to use react-boostrap for the modal? It isn't terrible, but it does have some issues with keyboard navigation and screen readers (for example, it puts the dialog itself into the natural tab order and it doesn't handle `shift+tab` very well). I can live with it if we don't want to bring in yet another package just for modals, but if this is a possibility then I will test some other react dialogs.

**Comment by bbsmooth** [comment]
@raditotev After looking into this further, I think that outer div with `role=dialog` is being added by the react-bootstrap dialog and so we'll just have to live with it. It probably isn't doing any harm.

Regarding the linter autofocus warning, as far as I'm concerned, the linter is wrong, at least with modal dialogs. When a dialog is opened there are several options for placing initial focus. You can place it on the dialog itself, on an interactive control (such as a button or input), or even on a non-focusable element (like a heading). And it is looking like the [`<dialog>` element will officially support the `autofocus` attribute](https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal) to allow initial focus to be placed on a specific element when the dialog is opened. In fact, it is going to be recommended that you always use `autofocus` when opening a modal dialog.

But if we are going to use react-bootstrap for the modal then I may recommend that we not use `autofocus` at all. So I would leave it out for now and we'll revisit after we have had time to do more testing.

**Comment by raditotev** [comment]
Thanks @bbsmooth. I have added the `aria-labelledby` and changed the title heading tag to `h1`. I have dropped the rest as suggested in you last comment. In my opinion if the team decides to move away from `react-bootstrap` for the modal and/ or any other elements, the work should be done in a separate issue.

**Comment by bbsmooth** [comment]
As far as accessibility goes (ignoring some issues with bootstrap-react itself), LGTM.

My only remaining question is whether we need two close buttons? Unless we add a link to this pop-up at some point then only keyboard users are going to use this. Seems a little redundant to have two close buttons in the tab order.

**Comment by raisedadead** [comment]
Yes - We are moving away from component libraries like bootstrap and derivatives, but it's going to be a bit. We are building our own minimal and specific component library with Tailwind (checkout the pinned GitHub issues, if interested) and we aim to make a11y a first-class citizen of the same.

As for the minor quirks we still have with the current state of the PR, I say we keep moving and land this in as "acceptable" for now. That way we can focus some more time on the component library should any of you be interested? :)

**Comment by raditotev** [comment]
@raisedadead I'll need help with this one. Frontend seems to be working but I get error when calling the server:
```
{"type":"danger","message":"flash.invalid-update-flag"}
```

**Comment by raisedadead** [comment]
As for the design, I think the message and the toggle are taking up too much real estate. Maybe we do not want to explain anything about how these work. Less is better if it's intuitive. We can use a detailed announcement post on the forum instead. Users are likely to discover the feature there.

I think a toggle (smaller in size) and a label "Enable Keyboard Shortcuts" should be more than plenty. But I will defer to other ideas if someone is interested in giving feedback too.

Functionality wise, once you have the same toggle in settings page we should be good to go.

File: api-server/src/server/boot/settings.js
Comment (raditotev) [reply api-server/src/server/boot/settings.js:236]: Yes

**Comment by raditotev** [comment]
@raisedadead this might be done. I think you can remove the `work in progress` label and let reviewers have their say.

File: client/src/client-only-routes/show-settings.tsx
Comment (nhcarrigan) [inline client/src/client-only-routes/show-settings.tsx:73]: This should use the new `setKeyboardShortcuts` saga I think.

**Comment by nhcarrigan** [comment]
So when I turn off keyboard shortcuts, it looks like that only disables the modal. `n` and `p` still switch challenges.

**Comment by raditotev** [comment]
Some of the e2e test are currently failing, because with the latest changes shortcuts are available only to logged users. From @raisedadead [comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/45530#issuecomment-1104700437) I understand this is the desired behaviour:
> The feature is aimed at our long-time and loyal users—rewarding them with something nice for sticking around in the community, and discovering it like an easter egg.

But on the other hand if this was open to unauthorised users up until know, do we really want to change that?
Depending on your comments there are 2 options:
* update tests, so that they include a login step in case we want to make shortcuts available to members only
* update `Hotkeys.tsx`, so that shortcuts are available to anyone and don't work only for members who explicitly turned them off

**Comment by nhcarrigan** [comment]
Given the initial accessibility concerns that @bbsmooth brought up, I'd vote for

> update tests, so that they include a login step in case we want to make shortcuts available to members only

That is, the keyboard shortcuts should be something a camper explicitly opts in to, rather than something that is on by default unless you sign up.

**Comment by raditotev** [comment]
@nhcarrigan with the latest changes, shortcuts for users are disabled by default. Also updated couple tests to reflect the changes.
In regards to `cypress/integration/learn/challenges/output.js`, the submit challenge shortcut works there because it's part of the editor.

**Comment by raditotev** [comment]
There are three failures in the [last test run](https://github.com/freeCodeCamp/freeCodeCamp/runs/6325809649?check_suite_focus=true) but none of them are related to changes in this branch. Could someone with permission start another run?

**Comment by raditotev** [comment]
@raisedadead is there anything else left to do as part of this PR?

**Comment by raisedadead** [comment]
Not to my knowledge, some one should look at this and merge as soon as we have resolved the ongoing issues with the platform.

**Comment by raisedadead** [comment]
@raditotev I have resolved the merge conflict and we will get this in soon. At some point, it would be ideal to mimic the "sound" feature's journey. That is we are moving to dedicated end-points. You do not have to do this just yet. Maybe a future PR.

Here is the reference of what I did: [`6338734` (#45530)](https://github.com/freeCodeCamp/freeCodeCamp/pull/45530/commits/6338734f091ed9c363d0efb158bd615d1ac8ad18)

cc @nhcarrigan

**Review by raisedadead — APPROVED** [review]

**Comment by nhcarrigan** [comment]
> That is we are moving to dedicated end-points. You do not have to do this just yet. Maybe a future PR

@raisedadead It looks like this PR creates the endpoint in the API but not in the client, which is exactly what we want - we can update the client in a separate PR once this is deployed. :)

**Review by naomi-lgbt — APPROVED** [review]
I am happy with this, and propose we get it in now - we can iterate further on it if desired in future PRs.

**Pull request merged**
